File: week_1/day_1_arrays/kadane_follow_up_questions/Can_this_be_applied_to_a_stream_of_numbers.py
class StreamingKadane:
    def __init__(self):
        self.curr_sum = None
        self.max_sum = None

    def add(self, num):
        if self.curr_sum is None:
            self.curr_sum = self.max_sum = num
        else:
            self.curr_sum = max(num, self.curr_sum + num)
            self.max_sum = max(self.max_sum, self.curr_sum)

        return self.max_sum


sk = StreamingKadane()
print(sk.add(-2))  # -2
print(sk.add(1))   # 1
print(sk.add(-3))  # 1
print(sk.add(4))   # 4
print(sk.add(-1))  # 4
print(sk.add(2))   # 5
print(sk.add(1))   # 6
print(sk.add(-5))  # 6
print(sk.add(4))   # 6

File: week_1/day_1_arrays/kadane_follow_up_questions/Can_this_be_applied_to_a_stream_of_numbers.js
class StreamingKadane {
    constructor() {
        this.currSum = null;
        this.maxSum = null;
    }

    add(num) {
        if (this.currSum === null) {
            this.currSum = this.maxSum = num;
        } else {
            this.currSum = Math.max(num, this.currSum + num);
            this.maxSum = Math.max(this.maxSum, this.currSum);
        }
        return this.maxSum;
    }
}

const sk = new StreamingKadane();

console.log(sk.add(-2)); // -2
console.log(sk.add(1));  // 1
console.log(sk.add(-3)); // 1
console.log(sk.add(4));  // 4
console.log(sk.add(-1)); // 4
console.log(sk.add(2));  // 5
console.log(sk.add(1));  // 6
console.log(sk.add(-5)); // 6
console.log(sk.add(4));  // 6
